Indexer fix for large articles

frankmayer opened the pull request:
Pull Request for Issue #22098 .

### Summary of Changes
This PR re-adds and optimizes chunking (for the VALUES clause) to the smart search indexer, allowing for larget articles to be saved correctly, and their terms to be correctly added to the index, again.
It also optimizes a condition check in the helper.

With a standard configuration of a PHP memoy limit at 128MB and the default Mysql HEAP limits for MEMORY tables these instructions should provide solid results.


### Testing Instructions

#### Introduction:

There are three scenarios. Each one involves saving a predefined article and getting an expected result.
The first scenario uses a text of ~32000 words. This one should be saved successfully. 
In the second scenario, we use a text of ~45000 words. This one will fail to save because of surpassing the finder_token table limits. 
Finally, in the third scenario we use a text of ~65000 words. This should cause a php fatal error, stating that the memory has been exhausted.


#### Preparation: 

- **Error reporting** should be set to **development mode**
- Enable the "Content - Smart Search" plugin (and the other Smart Search related) 
- Open a tab for a new article
- Open a tab for components/Smart Search

#### Scenarios:

1) Article with near maximum possible text
 a) On the "Smart Search" tab, click "Clear Index"
 b) Copy and paste the complete text (select all) from this [gist](https://gist.githubusercontent.com/frankmayer/f0864c40597a7f71d33da63ca18b7599/raw/35c992afa59fde43298c41e3a5889b4becab939a/lorem%2520x5) (213,4KiB unformatted text) into the article in the other tab
 c) Give it a title and "save" the article
 d) Expected result:  **Message** Article saved.
 e) On the "Smart Search" tab, click "Statistics" and check that it says "The indexed content on this site includes 11,995 terms across 1 links..." 

2) Article which exceeds the maximum entries for the `finder_token` table
 a) On the Smart Search tab, click "Clear Index"
 b) Copy and paste the complete text (select all) from this [gist](https://gist.githubusercontent.com/frankmayer/f0864c40597a7f71d33da63ca18b7599/raw/35c992afa59fde43298c41e3a5889b4becab939a/lorem%2520x7) (298,8KiB unformatted text) into the article in the other tab
 c) Save the article
 d) Expected result:  **Error** Save failed with the following error: The table '#__finder_tokens' is full
 e) On the "Smart Search" tab, click "Statistics" and check that it says "The indexed content on this site includes 0 terms across 1 links..."
**Remarks** 
If you get a message that the acrticle was saved, instead of the error, you probably have higher heap settings for the MEMORY tables in mysql. Still this result would count as expected.
If you get a php error showing that the memory is exhausted, you are probably running php with a limit of less than 128MB. If that happens, please copy/paste the error message into a comment of this issue.

3) Article which results in a fatal php error due to memory exhaustion
 a) On the Smart Search tab, click "Clear Index"
 b) Copy and paste the complete text (select all) from this [gist](https://gist.githubusercontent.com/frankmayer/f0864c40597a7f71d33da63ca18b7599/raw/35c992afa59fde43298c41e3a5889b4becab939a/lorem%2520x10) into the article in the other tab
 c) Save the article
 d) Expected result: a php error showing that the memory is exhausted
 e) On the "Smart Search" tab, click "Statistics" and check that it says "The indexed content on this site includes 0 terms across 1 links..."
**Remarks** 
If you get a message that the acrticle was saved, instead of the error, you probably have higher heap settings for the MEMORY tables in mysql. Still this result would count as expected.
If you get the error: The table '#__finder_tokens' is full, you might have a higher memory limit than 128MB on PHP.


### Documentation Changes Required
none

frankmayer committed:
Fix indexer memory exhaustion.

frankmayer committed:
Minor optimizations

ggppdk commented:
I have tested this item :white_check_mark: successfully on 8d9a9af439df561ee62c890042d52f12d71544d3<br /><br />This PR
addresses 3 things introduced recently

- re-adds chunking to avoid the SQL error due to large query length
- restores the peak memory usage
- restores a small slow-down when saving large articles

with PHP 7.2 + my old workstation 

Old code before PR #12253
memory_get_peak_usage(false) 70.0 MBs
memory_get_peak_usage(true)  12.0 MBs 
microtime(true) - $_SERVER["REQUEST_TIME_FLOAT"]   4.14 seconds (an average of repetitive runs)

Current code
memory_get_peak_usage(false)  107.6 MBs
memory_get_peak_usage(true)   42.0 MBs 
microtime(true) - $_SERVER["REQUEST_TIME_FLOAT"]   4.75 seconds (an average of repetitive runs)

This PR
memory_get_peak_usage(false) 70.0 MBs
memory_get_peak_usage(true)  12.0 MBs 
microtime(true) - $_SERVER["REQUEST_TIME_FLOAT"]   4.12 seconds (an average of repetitive runs)

<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/22599">issues.joomla.org/tracker/joomla-cms/22599</a>.</sub>

infograf768 commented on administrator/components/com_finder/helpers/indexer/helper.php:
```diff
@@ -327,9 +327,9 @@ public static function isCommon($token, $lang)
+		if ($default === null && $lang === '*')
 		{
-			$default = $default === null ? substr(self::getDefaultLanguage(), 0, 2) : $default;
+			$default = substr(self::getDefaultLanguage(), 0, 2);
```
Just remarked that this code (already present before this PR) will not work when language has a langcode of type `xxx-XX`. (We have a few in J).
Would we ever get common terms for these languages? No idea.
Lang concerned are ckb-IQ, prs-AF, srp-ME as of 3.9.0

The code to modify this would be something like
```
						$tag = strlen(self::getDefaultLanguage());

						if ($tag == 5)
						{
							$default = substr(self::getDefaultLanguage(), 0, 2);
						}
						elseif ($tag == 6)
						{
							$default  = substr(self::getDefaultLanguage(), 0, 3);
						}
```

frankmayer replied:
@infograf768, thanks for the remark. Noted. I think that these kinds of functions also have to be changed at several other language-related places, if I am not mistaken. Lets address this in a separate PR, ok?

infograf768 replied:
We use that kind of code in each core component admin/views modal.php when multilingual is implemented to fill up the `data-language` data attribute when selecting an item through javascript `JSelect`.  
**I doubt btw of its usefulness in these cases as it remains a select only markup and is never used for display.**

Not sure we need it elsewhere where we need for links or the full lang code or the sef prefix, which are the only correct way to differentiate between same language but different country code .
In core we never use in frontend the data attributes as far as I know and if we were using it **we should use the full lang code.**

In fact, I think that the only place where it would be useful atm is here in finder where we do not differentiate between en-GB and en-US for common words.

frankmayer replied:
Oh, ok. Then I'll make the change and update this PR, to get this in.

ggppdk commented:
About the testing article text given here
with old and new code i am getting 
**Save failed with the following error: The table '#__finder_tokens' is full**

I had to increase
max_heap_table_size = 64M
to 
max_heap_table_size = 128M     (or more)

and restart MySql
to be able to save the article

ggppdk commented:
aaah ,
my last comment was wrong

article with save with max_heap_table_size = 64M,
i thought i was testing with 32K words sample but i had pasted in the editor a bigger sample !

frankmayer commented:
> 
> Old code before PR #12253
> 
> * 68.9 MBs / 10.0 MBs / 11.07 seconds
> 
> Current code
> 
> * 106.5 MBs / 44.0 MBs / 11.28 seconds
> 
> This PR
> 
> * 68.9 MBs / 10.0 MBs /  11.06 seconds
> 
Thanks for testing this, George. Though I don't know what caused the memory consumtion to rise.. (I believe the "not chunking", without having tested it, though). I ran some tests back then, but cannot seem to find any notes on that. Should have kept them.

frankmayer committed:
Implemented support for 3-character language codes `xxx-xx` in addition to `xx-xx` as remarked here: https://github.com/joomla/joomla-cms/pull/22599#pullrequestreview-164486405
The use of `strstr` was chosen because it is less code and twice as fast as having conditions for 2 or 3 letters.

frankmayer commented:
@ggppdk I updated this PR according to @infograf768 's remark above. I used `strstr()` though, for readability and performance.

Quy commented on administrator/components/com_finder/helpers/indexer/indexer.php:
```diff
@@ -510,30 +513,44 @@ protected function addTokensToDb($tokens, $context = '')
 		$db = $this->db;
 
-		$query = clone $this->addTokensToDbQueryTemplate;
 
```
Remove extra blank line.

frankmayer replied:
Why? It would look like the cloning of the query is part of getting the database object. Like this it is easier to read. 
Also, why don't you give @ggppdk a break, he already tested this twice :smile:

frankmayer replied:
ok, thanks, missed that...

frankmayer committed:
Remove extra blank line

frankmayer commented:
@ggppdk @Quy remarked an extra blank line, so I had to commit and push again. Could you pls confirm a successful test? (Nothing changed, so just mark it)

ggppdk commented:
I have tested this item :white_check_mark: successfully on be81f0e76d577831f77e3ef43c9e2a9b50cabea8<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/22599">issues.joomla.org/tracker/joomla-cms/22599</a>.</sub>

dagalumin commented:
I have tested this item :white_check_mark: successfully on be81f0e76d577831f77e3ef43c9e2a9b50cabea8<br /><br />This new PR works great, bringing back the ability to save large articles as on 3.8.11 version of the indexer, plus adding much better performance:
on the not-so-shiny server of a client with Joomla 3.8.11 version of the indexer we had a limit of about 25k words per article, this PR extended it to 40k with the same 500mb memory limit.
Thank you so much guys for the great work!<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/22599">issues.joomla.org/tracker/joomla-cms/22599</a>.</sub>

Quy commented:
RTC<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/22599">issues.joomla.org/tracker/joomla-cms/22599</a>.</sub>
